TumbleBitSetup/PoupardStern.cs

```diff
@@ -0,0 +1,263 @@
+        /// </summary>
+        /// <param name="pubKey">Public key used</param>
+        /// <param name="ps">public string specified in the setup</param>
+        /// <param name="i">index i</param>
+        /// <param name="k">Security parameter specified in the setup.</param>
+        /// <param name="keyLength">The size of the RSA key in bits</param>
+        /// <returns></returns>
+        internal static BigInteger SampleFromZnStar(RsaPubKey pubKey, byte[] psBytes, int i, int k, int keyLength)
+        {
+            BigInteger Modulus = pubKey._pubKey.Modulus;
+            GetK(k, out int BigK); // I think it would be more efficient to calculate K once outside this function.
+            int j = 2;
+            // Octet Length of i
+            int iLen = Utils.GetByteLength(BigK);
+            // Byte representation of i
+            var EI = Utils.I2OSP(i, iLen);
+            // ASN.1 encoding of the PublicKey
+            var keyBytes = pubKey.ToBytes();
+            // Combine the octet string
+            var combined = Utils.Combine(keyBytes, Utils.Combine(psBytes,EI));
+            for (;;)
+            {
+                // OctetLength of j
+                var jLen = Utils.GetOctetLen(j);
+                // Byte representation of j
+                var EJ = Utils.I2OSP(j, jLen);
+                // Combine EJ with the rest of the string
+                combined = Utils.Combine(combined, EJ);
+                // Pass the bytes to H_1
+                byte[] ER = Utils.MGF1_SHA256(combined, keyLength);
+                // Convert from Bytes to BigInteger
+                BigInteger z_i = Utils.OS2IP(ER);
+                // Check if the output is larger or equal to N OR GCD(z_i, N) != 1
+                if (z_i.CompareTo(Modulus) >= 0 || !(z_i.Gcd(Modulus).Equals(BigInteger.One)))
+                {
+                    j++;
+                    continue;
+                }
+                return z_i;
+            }
+        }
+
+        /// <summary>
+        /// Calculates the value of w as specified in the setup.
+        /// </summary>
+        /// <param name="pubKey">Public key used</param>
+        /// <param name="ps">public string specified in the setup</param>
+        /// <param name="xValues"> List of x_i values</param>
+        /// <param name="k">Security parameter as specified in the setup.</param>
+        /// <param name="keyLength">The size of the RSA key in bits</param>
+        internal static void GetW(RsaPubKey pubKey, byte[] psBytes, BigInteger[] xValues, int k, int keyLength, out BigInteger w)
+        {
+            if (xValues == null)
+                throw new ArgumentNullException(nameof(xValues));
+
+            var BigK = xValues.Length;
+
+            // ASN.1 encoding of the PublicKey
+            var keyBytes = pubKey.ToBytes();
+
+            // Encoding of the x_0
+            var ExLen = Utils.GetByteLength(keyLength);
+            var Ex_0 = Utils.I2OSP(xValues[0], ExLen);
```

Comment:
a057d4f1cc53cd08d5c0ba5c9246bd38d97f1c58
So I reviwed this part and made it a bit cleaner and shorter. I couldn't initialize ExComb to null because when it gets passed to "Combine" it throws a "NullReferenceException", so I just initialized to an empty array.
Let me know what do you think